### PR TITLE
Backport checkquote tool and PCR outputs for quote tool to 3.X

### PR DESCRIPTION
**Pull request by jetwhiz**
Backporting of features added to master branch to 3.X: 
* Add new tpm2_checkquote tool for checking quotes
  * Pull shared functionality from tpm2_pcrlist into pcr library
  * is_pcr_select_bit_set moved into tpm2_util library
  * Add new functionality to openssl, pcr and util libraries
* The tpm2_quote tool can now output PCR hash lists

New: 
* Add attestation test to check that the full attestation process works

### CODE REVIEW

File: lib/tpm2_util.c [hunk 1]
```diff
@@ -41,6 +41,157 @@
 #include "tpm2_tool.h"
 #include "tpm2_util.h"
 
+#define FILE_PREFIX "file:"
+#define FILE_PREFIX_LEN (sizeof(FILE_PREFIX) - 1)
```
Comment (williamcroberts) [inline lib/tpm2_util.c]: I don't think you need this hunk

File: man/tpm2_quote.1.md [hunk 2]
```diff
@@ -41,18 +41,25 @@
 
   * **-m**, **--message**:
 
-    message output file, records the quote message that makes up the data that
+    Message output file, records the quote message that makes up the data that
```
Comment (williamcroberts) [inline man/tpm2_quote.1.md]: separate patch for these to fixups in tpm2_quote.md.1

**Comment by jetwhiz** [comment]
> I noticed the test_attestation.sh script and the verifyquote tool isn't on master, are you planning on porting that to master?

The verify quote tool should be on master, but we put it under the `tools/misc` directory since it does not use the TPM at all -- [master/tools/misc/tpm2_checkquote.c](https://github.com/tpm2-software/tpm2-tools/blob/master/tools/misc/tpm2_checkquote.c).  

Regarding the attestation test, we have one ready for master but it does not pass right now due to authentication issues on the master branch (as discussed on the mailing list: https://lists.01.org/pipermail/tpm2/2019-February/001046.html).  If you'd, like we can submit this with authentication parts stripped out, and then we can add it back after these issues are taken care of?

**Comment by williamcroberts** [comment]
Looks like this bug is filed for 4.0: https://github.com/tpm2-software/tpm2-tools/issues/1453

File: man/tpm2_quote.1.md (hunk 2 above)
Comment (jetwhiz) [reply man/tpm2_quote.1.md]: Should this be a separate PR or just in its own patch/commit?

Comment (williamcroberts) [reply man/tpm2_quote.1.md]: Its own commit on this PR is 100% OK.

File: lib/tpm2_util.c (hunk 1 above)
Comment (jetwhiz) [reply lib/tpm2_util.c]: Good catch ... I forgot to strip that out when I ported from master

**Comment by jetwhiz** [comment]
I'll keep an eye on that issue and push the attestation tests when the mentioned fix gets merged.  We're also looking to start tracking master for Keylime as well, so we'll be keeping an eye on that issue regardless.

File: lib/pcr.h [hunk 3]
```diff
@@ -35,8 +35,35 @@
 
 #include <tss2/tss2_sys.h>
 
+typedef struct tpm2_algorithm tpm2_algorithm;
+struct tpm2_algorithm {
+    int count;
+    TPMI_ALG_HASH alg[8]; //XXX Why 8?
```
Comment (williamcroberts) [inline lib/pcr.h]: You should be backporting from the tip of master so you pick up commits like 1bdaef80a8ad7 and 785c7cbc1927

Comment (jetwhiz) [reply lib/pcr.h]: I just noticed that [82d987](https://github.com/tpm2-software/tpm2-tools/commit/82d987a8e75ebefec4cbf1c2bd3fda87c2aa5e34) hasn't been backported.  Should I include this commit?

Comment (williamcroberts) [reply lib/pcr.h]: Yep

Comment (jetwhiz) [reply lib/pcr.h]: Ok, I cherry picked that commit.  The latest should incorporate everything related to this PR from the tip of master.

**Pull request merged**